### PR TITLE
GH-15054: [Python] Finalize S3 on pytest exit to prevent shutdown crashes

### LINKED ISSUE #15054: [CI][Python] wheel-manylinux2014-* sometimes crashed on pytest exit
### Describe the bug, including details regarding any error messages, version, and platform.

wheel-manylinux2014-cp38-amd64 2022-12-20 nighty:

https://github.com/ursacomputing/crossbow/actions/runs/3738813751/jobs/6345300403#step:9:775

```text
= 4294 passed, 382 skipped, 18 xfailed, 2 xpassed, 4 warnings in 522.05s (0:08:42) =
/arrow/ci/scripts/python_wheel_unix_test.sh: line 95:    79 Segmentation fault      (core dumped) python -m pytest -r s --pyargs pyarrow
```

wheel-manylinux2014-cp39-arm64 2022-12-20 nighty:

https://app.travis-ci.com/github/ursacomputing/crossbow/builds/259026563#L5708

```text
= 4324 passed, 352 skipped, 18 xfailed, 2 xpassed, 1 warning in 141.20s (0:02:21) =
/arrow/ci/scripts/python_wheel_unix_test.sh: line 95:    15 Segmentation fault      (core dumped) python -m pytest -r s --pyargs pyarrow
```

### Component(s)

Continuous Integration, Packaging, Python

### PR DESCRIPTION
**Pull request by westonpace**
For some reason destroying an S3Client at shutdown can cause a crash in our tests.  It appears to be because S3's logging is already shut down.  I wouldn't think this would happen because I would think the linker would tear down S3 shared libraries AFTER it tears down Arrow shared libraries.

However, we can also finalize S3 in conftest.py to prevent this crash.

It would still be good to understand the root cause.  Otherwise we need to advise our users to call finalize_s3.
* Closes: #15054

### CODE REVIEW

**Comment by lidavidm** [comment]
> because I would think the linker would tear down S3 shared libraries AFTER it tears down Arrow shared libraries.

In a wheel though, we statically link everything, and then destruction order is undefined as usual?

File: python/pyarrow/conftest.py [hunk 1]
```diff
@@ -265,3 +265,13 @@ def add_fs(doctest_namespace, request, tmp_path):
         doctest_namespace["local_path"] = str(tmp_path)
         doctest_namespace["path"] = str(path)
     yield
+
```
Comment (raulcd) [inline python/pyarrow/conftest.py:268]: I think the linter error is only because it expects two empty lines here.
```suggestion


```

**Comment by westonpace** [comment]
> In a wheel though, we statically link everything, and then destruction order is undefined as usual?

Yes, I think that could explain it.  Still, this fix would require pyarrow users to know about and remember calling the finalize S3 method.  I'm going to pause this idea and make an attempt at an arrow::Shutdown.

**Comment by westonpace** [comment]
Closing in favor of #33858

**Pull request closed without merging**